lib/jekyll/readers/theme_data_reader.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+
+module Jekyll
+  class ThemeDataReader < Jekyll::DataReader
+    attr_reader :site, :content
+
+    def read(dir)
+      return unless site.theme&.data_path
+
+      base = site.in_theme_dir(dir)
+      read_data_to(base, @content)
+      Jekyll::Utils.deep_merge_hashes(@content, @site.data)
+    end
+
+    def read_data_to(dir, data)
+      return unless File.directory?(dir) && !@entry_filter.symlink?(dir)
+
+      entries = Dir.chdir(dir) do
+        Dir["*.{yaml,yml,json,csv,tsv}"]
+      end
+
+      entries.each do |entry|
+        path = @site.in_theme_dir(dir, entry)
```

Comment:
This is the main difference between the superclass and this class, so maybe we can use `@source_dir` here instead, or use a `@source_dir_proc` that allows us to pass the method `in_source_dir()` or `in_theme_dir()` around.